pkg/controller/node/node_controller.go

```diff
@@ -128,11 +153,235 @@ func (ctrl *Controller) Run(workers int, stopCh <-chan struct{}) {
+			}
+		}
+	}
+	return errors
+}
+
+// makeMastersUnSchedulable makes all the masters in the cluster unschedulable
+func (ctrl *Controller) makeMastersUnSchedulable(currentMasters []*corev1.Node) []error {
+	errors := []error{}
+	for _, node := range currentMasters {
+		if !CheckMasterIsAlreadySchedulable(node) {
+			continue
+		} else {
+			if err := ctrl.makeMasterNodeUnSchedulable(node); err != nil {
+				errors = append(errors, fmt.Errorf("failed making node %v schedulable with error %v", node.Name, err))
+			}
+		}
+	}
+	return errors
+
+}
+
+// makeMasterNodeUnSchedulable makes master node unschedulable by removing worker label and adding `NoSchedule`
+// master taint to the master node
+func (ctrl *Controller) makeMasterNodeUnSchedulable(node *corev1.Node) error {
+	_, err := updateNodeRetry(ctrl.kubeClient.CoreV1().Nodes(), ctrl.nodeLister, node.Name, func(node *corev1.Node) {
+		// Remove worker label
+		newLabels := node.Labels
+		if _, hasWorkerLabel := newLabels[workerLabel]; hasWorkerLabel {
+			delete(newLabels, workerLabel)
+		}
+		node.Labels = newLabels
+		// Add master taint
+		newTaints := node.Spec.Taints
+		MasterUnSchedulableTaint := corev1.Taint{Key: masterLabel, Effect: corev1.TaintEffectNoSchedule}
+		newTaints = append(newTaints, MasterUnSchedulableTaint)
+		node.Spec.Taints = newTaints
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// CheckMasterIsAlreadySchedulable checks if the given node has a worker label and doesn't have NoSchedule master
+// taint
+func CheckMasterIsAlreadySchedulable(node *corev1.Node) bool {
+	_, hasWorkerLabel := node.Labels[workerLabel]
+	hasMasterTaint := false
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == masterLabel && taint.Effect == corev1.TaintEffectNoSchedule {
+			hasMasterTaint = true
+		}
+	}
+	return hasWorkerLabel && !hasMasterTaint
+}
+
+// makeMasterNodeSchedulable makes master node schedulable by removing NoSchedule master taint and
+// adding worker label
+func (ctrl *Controller) makeMasterNodeSchedulable(node *corev1.Node) error {
+	_, err := updateNodeRetry(ctrl.kubeClient.CoreV1().Nodes(), ctrl.nodeLister, node.Name, func(node *corev1.Node) {
+		// Add worker label
+		newLabels := node.Labels
+		if _, hasWorkerLabel := newLabels[workerLabel]; !hasWorkerLabel {
+			newLabels[workerLabel] = ""
+		}
+		node.Labels = newLabels
+		// Remove master taint
+		newTaints := []corev1.Taint{}
+		for _, t := range node.Spec.Taints {
+			if t.Key == masterLabel && t.Effect == corev1.TaintEffectNoSchedule {
+				continue
+			}
+			newTaints = append(newTaints, t)
+		}
+		node.Spec.Taints = newTaints
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// This is duplicated code and is copied from package daemon. Perhaps a more neutral place for this exists. If not can we
+// create one?
+func updateNodeRetry(client corev1client.NodeInterface, lister corev1lister.NodeLister, nodeName string, f func(*corev1.Node)) (*corev1.Node, error) {
```

Comment:
ok, let's move this to `./internal/kubeutils`(?)

Reply:
Ohh, I just created a top level utils package with `node_utils.go`. I don't have a strong preference. I can move it `./internal/kubeutils`. Do you want me to do that?